hishel/_utils.py

```diff
@@ -33,17 +33,37 @@ def normalized_url(url: tp.Union[httpcore.URL, str, bytes]) -> str:
     assert False, "Invalid type for `normalized_url`"  # pragma: no cover
 
 
-def generate_key(request: httpcore.Request) -> str:
+def sync_generate_body_hash(request: httpcore.Request) -> str:
+    if isinstance(request.stream, tp.Iterable):
+        body_hash = blake2b(digest_size=16)
+        for chunk in request.stream:
+            body_hash.update(chunk)
+    elif isinstance(request.stream, tp.AsyncIterable):
+        raise TypeError("AsyncIterable streams are not supported for request bodies in the sync client.")
+    return body_hash.hexdigest()
+
+
+async def async_generate_body_hash(request: httpcore.Request) -> str:
+    if isinstance(request.stream, tp.AsyncIterable):
+        body_hash = blake2b(digest_size=16)
+        async for chunk in request.stream:
+            body_hash.update(chunk)
+    elif isinstance(request.stream, tp.Iterable):
+        body_hash = blake2b(digest_size=16)
+        for chunk in request.stream:
+            body_hash.update(chunk)
```

Comment:
Let's keep the async and sync API's same and raise an exception here like we do in async version.